File: sgd.py
from optimizer import Optimizer
from tensor import Tensor

class SGD(Optimizer):
    def __init__(self, params, lr=0.001):
        super().__init__(params)
        self.lr = lr
    def step(self):
        for _, param in enumerate(self.params):
            grad = Tensor.from_numpy(param.grad.data, dtype=param.grad.dtype)
            param.data = param.data - grad * self.lr
    def zero_grad(self):
        for param in self.params:
            param.grad = None

File: torchlib/src/optimizer/sgd.rs
use crate::tensor::{tensor::TensorId, tensor_factory::TensorFactory};

use super::optimizer::Optimizer;

pub struct SGD {
    params: Vec<TensorId>,
    lr: f64,
}

impl Optimizer for SGD {
    fn step(&mut self, factory: &mut TensorFactory) {
        self.params.iter().for_each(|p| {
            let t = factory.get_mut(p).unwrap();
            let grad = t.grad.as_ref().unwrap();
            let data = t.cached_data.as_ref().unwrap();
            t.cached_data = Some(data - grad * self.lr);
        });
    }

    fn reset_grad(&mut self, factory: &mut TensorFactory) {
        self.params.iter().for_each(|p| {
            factory.get_mut(p).unwrap().grad = None;
        })
    }

    fn update_lr(&mut self, lr: Vec<f64>) {
        self.lr = lr[0];
    }

    fn lr(&self) -> Vec<f64> {
        vec![self.lr]
    }
}

impl SGD {
    pub fn new(params: Vec<TensorId>, lr: f64) -> Self {
        Self { params, lr }
    }
}
